### PR TITLE
[Datasets] Arrow 7.0.0+ Support: Use Arrow IPC format for pickling Arrow data to circumvent slice view buffer truncation bug.

### LINKED ISSUE #22310: [Dataset] ray failed to serialize pyarrow7.0.0 Tables
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

See https://github.com/ray-project/ray/pull/22253#issuecomment-1034742274 and https://github.com/ray-project/ray/issues/22177 for more context.

TL;DR, when serialize pyarrow7.0.0 Tables, the serialization has wrong data size (as big as 200GBs). 

### Versions / Dependencies

pyarrow 7.0.0

### Reproduction script

python ray/release/nightly_tests/dataset/ray_sgd_runner.py --num-epochs 1 --smoke-test
with pyarrow 7.0.0 installed

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR registers a custom serializer for Array arrays, chunked arrays, record batches, and tables that works around an [Arrow serialization bug](https://issues.apache.org/jira/browse/ARROW-10739) that serializes the entire underlying data buffer when serializing zero-copy slice views. The custom serializer uses the Arrow IPC format as the underlying pickled representation.

This is an alternative to #22891, which recursively traverses the buffers in the Arrow arrays and manually truncates them if there was an array-level offset. This alternative route, which piggybacks off of the Arrow IPC format, has the following pros and cons:

### Pros

1. **Less complex:** We're delegating the buffer truncation to Arrow's IPC machinery.
2. **Automatic type coverage:** We inherit the Arrow array type coverage of Arrow's IPC format, which is thorough and is covered in upstream Arrow's CI.
3. **Easier maintenance:** New Arrow array types or changing memory layouts will be handled upstream by Arrow.
4. **Faster serialization:** All type traversals happen in Arrow's C++ code rather than in our recursive buffer traversal + copying, which should be faster.

### Cons

1. **Serialization bloat:** Arrow's IPC format is geared towards RecordBatches, so Arrow arrays will be wrapped with a redundant `RecordBatch` when serialized, which adds serialization bloat of ~230 bytes to each Arrow array. Thankfully, Datasets primarily serializes Arrow Tables, which typically serialize more compactly in the IPC format than the current Arrow pickled format, so we directly register a custom serializer for Arrow Tables to bypass this con.
2. **Less control of format:** if Arrow's IPC format doesn't support a particular type, we'll need to add workarounds for it. This isn't likely to happen, though, since Arrow is invested in full coverage.

## Related issue number

Closes https://github.com/ray-project/ray/issues/22310

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset.py
Comment (jianoaix) [inline python/ray/data/dataset.py:440]: nit: ``fn``

File: python/ray/data/_internal/arrow_serialization.py
Comment (jianoaix) [inline python/ray/data/_internal/arrow_serialization.py:154]: In which case do we expect this? The pyarrow is a required install when installing Ray.

File: python/ray/data/_internal/arrow_serialization.py
Comment (jianoaix) [inline python/ray/data/_internal/arrow_serialization.py:245]: This seems to need memory allocation and no longer zero-copy. Can it serialize each chunk to a output without moving chunks into contiguous region?

**Review by jianoaix — COMMENTED** [review]
Looks massively simpler than previous PR!

File: python/ray/data/_internal/arrow_serialization.py
Comment (clarkzinzow) [reply python/ray/data/_internal/arrow_serialization.py]: Yep I had a TODO for finding a way around this, turns out the `RecordBatchStreamWriter.write_table()` API will write each underlying horizontal set of column chunks as a record batch, so it will be zero-copy. I just ported the `ChunkedArray` serialization path to that API!

File: python/ray/data/_internal/shuffle_and_partition.py
Comment (c21) [inline python/ray/data/_internal/shuffle_and_partition.py:59]: sorry if I miss anything, why it's safe to remove `copy=True` here?

File: python/ray/data/_internal/arrow_serialization.py
Comment (clarkzinzow) [reply python/ray/data/_internal/arrow_serialization.py]: I moved this check to happen once, but we need this check when registering these serializers because `pyarrow` is not a dependency when installing the minimal Ray package via `pip install ray`, it's part of the `ray[data]` extra requirements that would be installed with `pip install ray[data]`: https://github.com/ray-project/ray/blob/3c4302906fd0d3958ccf2588ad37dd5b9912b1d9/python/setup.py#L226

So we need this serializer to short-circuit when `pyarrow` is not installed, since this serializer registration code will happen on `import ray`.

File: python/ray/data/_internal/shuffle_and_partition.py
Comment (clarkzinzow) [reply python/ray/data/_internal/shuffle_and_partition.py:59]: This `copy=True` here (and elsewhere in the PR) was to avoid a smaller-scope version of this Arrow serialization bug in `pyarrow <= 6.0.1` that we observed a while back, where the zero-copy slice views produced by `copy=False` would result in the entire `block` buffer to be sent for every slice. To circumvent this, we eagerly copied the block slices.

With this Arrow serialization hook, this copy (i.e. "buffer truncation", where the zero-copy slice view is turned into a truncated copy of the corresponding slice of the buffer) will happen automatically at serialization time, so we don't need to explicitly copy the slices here.

**Comment by c21** [comment]
Looks good to me overall. @jianoaix - could you help take another look? Also please resolve conflict, thanks.

**Comment by clarkzinzow** [comment]
Thanks for taking a look @c21!

>Also please resolve conflict, thanks.

Yep, I'm just waiting for https://github.com/ray-project/ray/pull/29071 to be merged since the original PR being reverted is the source of the conflicts, and should be resolved once the unreverting PR is merged.

**Review by jianoaix — APPROVED** [review]
Well done!

**Comment by clarkzinzow** [comment]
@ericl @richardliaw @edoakes Need a stamp from a code owner.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
It may be a good idea to flag flip this on in a separate PR to avoid having to revert the entire thing.

**Comment by scv119** [comment]
@clarkzinzow looks this PR might have broken Windows test 
image

**Pull request merged**
